Feature: Attachment type

avisionh opened the pull request:
# Summary
This branch implements the ability to compute the percentage of PDF attachment on GOV.UK pages. It follows a [request](https://trello.com/c/Gfh9LCt4) made by the Accessibility Team for this so they can audit the progress made to either remove PDF attachments or provide accessible versions of them.

In particular, for each page, have counts of the attachment types.
- Especially include PDF, HTML and ODS.

Include extra info including:
- Slug
- Publishing organisation 
- Doc type
- Created at

Need this for pages created for all time and for pages created in last 5/6 months (Jul - Nov 2020). We handle this by just providing all pages created across time.

## Check
- [ ] Code works.
- [ ] The report gives what is asked for in the request.

chao-xian reviewed (commented):
Can we move the files in the /notebooks directory into the helpers directory. It also feels like they should have tests too.

chao-xian commented on README.md:
```diff
@@ -73,6 +73,13 @@ substituting `<REPORT CONFIG FILENAME>` for the report config you wish to use fr
+For some reports, some post-processing needs to be done before sharing with the Accessibility team. Where a report
+requires this, then you will find it under `notebooks/<report_name>_postprocess/py`.
+
+For example, if you want to create the `src/report_generators/attachment_type_report_generator.py` report, then run this
```
You could probably use the config file and add in a conditional into `run_accessibility_reports.py`. Also means we don't create multiple ways of generating reports.

avisionh replied:
Yes that would be the nicer way to do it but since we need to do some post-processing on the entire `csv` rather than line-by-line, I'm not sure if it can fit neatly within the `run_accessibility_reports.py` framework.

chao-xian replied:
Yes, sorry I was being vague and unclear: we could add to the `__main__` to run the post-process if the option in the config file for that report was set to run that report. Or would that not work? That way the documentation and ways of running it can be consistent for the user.

chao-xian replied:
Outside of the loops

avisionh replied:
Makes sense, I've added it now but not sure if I added it correctly. Will appreciate your help. 🙂 Unfortunately, I can't test because I need to download the mirror and since I'm using my personal macbook, I have no memory for it (serves me right for buying the lowest memory version!).

chao-xian replied:
Ah I don't have the files either... it'll take a while to download. I'll see if I can kick this off.

chao-xian commented:
@avisionh sorry I took so long to take a look at this. I've had a quick look only atm, and it looks pretty good! Just a couple of things atm. Will take a deeper look when I get a chance, but tbh I think the code is fine otherwise.

avisionh commented on src/utils/constants.py:
```diff
@@ -26,6 +26,7 @@
                       'updated_at',
                       'withdrawn_at']
 
+# from: https://www.gov.uk/guidance/how-to-publish-on-gov-uk/creating-and-updating-pages
```
Yup, done!

avisionh commented:
> Can we move the files in the /notebooks directory into the helpers directory. It also feels like they should have tests too.

Moved the files in the `/notebooks` directory into the `src/report_generators` directory so that they are better located now. Regarding tests, I've already added tests in the functions for the post-processing scripts. If you mean add a test for the whole script, that's a bit more difficult because it's currently being run line-by-line. To add test to the whole script, I will probably have to add tests to each part like the loops.

chao-xian commented:
@avisionh thanks for adding the tests for the post processing, that was an area that looked pretty new to me so was more keen to get tests for. Generally I follow this way of adding tests: thoroughly testing any part that gets then processes/manipulates data. Testing the whole process more broadly if possible is good practise too, but at least the unit test coverage is what I'd aim for but it's not a priority for this atm. If we come to find we're adding more and more changes to this in future we will definitely need to.

chao-xian reviewed (approved):
Nice work @avisionh! Thank you for doing this. Great stuff.
